### ISSUE TITLE
What is the point of calling ngx_http_modsecurity_pcre_malloc_init?

### ISSUE THREAD

**Issue by dennus** [SpiderLabs/ModSecurity-nginx#66]
Hello! Can you explain, why using of ngx_http_modsecurity_pcre_malloc_init before any calling of msc_proccess_* is needed? I don't found any additional operation with memory in examples of libmodsecurity usage, in apache connector and somewhere else, only in a Nginx connector. I see, that a typical operation in ModSecurity 2.9, but I can't understand, what does it mean in Modsecurity-Nginx? I try to comment this lines in all phase processing and don't see any changes in behavior.

But, if before this changes Nginx workers are overhead in memory usage after starting 500 parallel client thread, after my changes memory usage grow slower and Nginx can parallel handle more then 3000 client.

**Comment by zimmerle** [comment]
Hi @dennus,

nginx has it own way to handle the memory used by pcre. It hooks the pcre allocation and de-allocation methods to use memory in a smart manner, avoiding memory fragmentation. Turns out that we don't have the guarantee that this approach will be consistent out of the main nginx process. To avoid any kind of random crashes, we hook the allocation methods back to malloc (and free), giving it back on its original format just before handles the control over nginx.

That is not something special for ModSecurity, others nginx modules adhibit similar logic while handling this scenario, as example of Lua:
https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_pcrefix.c

I hope I have answered your question ;)
